### ISSUE TITLE
[BUG] The output of BitLinear is quite abnormal

### ISSUE THREAD

**Issue by Jiangxg** [kyegomez/BitNet#35]
**Describe the bug**
I print the mean and variance of the tensor y in example.py.
Its mean and variance are abnormal, as follows:

> mean and var of BitLinear output:
-0.567935049533844
1149.9969482421875

To make sure, I print the mean and variance of outputs from Linear and BitLinear, simutaneously.

> mean and var of Linear output:
0.012186492793262005
0.33256232738494873
mean and var of BitLinear output:
0.9070871472358704
992.69384765625

I believe there are mistakes in the implementation of BitLinear in bitnet/bitlinear.py.

**To Reproduce**
1. print the mean and variance of y in example.py
2. insert `output_linear = torch.nn.functional.linear(x, self.weight, self.bias)` in bitnet/bitlinear.py line 129. Then print the mean and variance of output_linear

**Comment by suzuke** [comment]
The implementation of this binear is completely wrong, not only does it not follow the process outlined in the Bitnet paper, but it also misunderstands all the computational principles. I don't understand why it still receives so many stars.

**Comment by suzuke** [comment]
Gemma, beta, and alpha are calculated using weights and input before quantization. These parameters are then utilized for weights binarization and input quantization. The binarized weights and quantized input undergo linear operations to produce the output, which is then dequantized using the previously calculated gemma, beta. It's not meaningful to calculate gemma and beta separately for quantization and dequantization stages, and even the implementation of grouping here is entirely nonsensical.

**Comment by 2020zyc** [comment]
> Gemma, beta, and alpha are calculated using weights and input before quantization. These parameters are then utilized for weights binarization and input quantization. The binarized weights and quantized input undergo linear operations to produce the output, which is then dequantized using the previously calculated gemma, beta. It's not meaningful to calculate gemma and beta separately for quantization and dequantization stages, and even the implementation of grouping here is entirely nonsensical.

hi, I don't understand what u say. Could u tell more?
The code just calculates the gamma/beta in quantization stage dynamically, then uses the two statistics to dequant activation. 
No extra calculation of gamma/beta in dequantization stage.
You of course can take the previous calculation out of the quantization stage, but still need dynamically get the gamma/beta.

**Comment by 2020zyc** [comment]
> Gemma, beta, and alpha are calculated using weights and input before quantization. These parameters are then utilized for weights binarization and input quantization. The binarized weights and quantized input undergo linear operations to produce the output, which is then dequantized using the previously calculated gemma, beta. It's not meaningful to calculate gemma and beta separately for quantization and dequantization stages, and even the implementation of grouping here is entirely nonsensical.

Another implementation is [BIT-Transformers](https://github.com/Beomi/BitNet-Transformers/tree/main). I don't know how its BitLinear works, especially the forward function. Could u understand this code? Thanks

[forward](https://github.com/Beomi/BitNet-Transformers/blob/6198f9cd9d37ac45e37fa9f8689a289dc0f724e0/bitnet_llama/module.py#L199)
![image](https://github.com/kyegomez/BitNet/assets/7539692/cb737bc4-81f3-40ef-a2cc-52d7a0bbcdc1)

**Comment by suzuke** [comment]
The issues I mentioned have been addressed in the commit 6cdb2ea998e843b454f2fbaaef73bc6bf92c305f.

**Comment by Jiangxg** [comment]
> The issues I mentioned have been addressed in the commit [6cdb2ea](https://github.com/kyegomez/BitNet/commit/6cdb2ea998e843b454f2fbaaef73bc6bf92c305f).

Yes, most of the problem has been addressed. Still got a bug in the implementation of grouping. I am working on that.
